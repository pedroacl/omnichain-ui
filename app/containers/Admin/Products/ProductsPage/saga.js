
import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import { LOAD_PRODUCTS, LOAD_PRODUCT } from './constants';
import { productLoaded, productsLoaded, errorLoadingProducts } from './actions';
import axios from 'axios';

import request from 'utils/request';

/**
 * Worker sagas
 */
export function* loadProducts(action) {
  try {
    const requestURL = 'products';
    // const params = { companyAddress: action.payload };
    // const products = yield call(() => axios.get('products', { params }).then(res => res.data));
    const products = [
      {
        name: 'Test Product 1',
        description: 'Test Product 1 description',
        currentStage: 'Test Product 1 current stage',
      },
      {
        name: 'Test Product 2',
        description: 'Test Product 2 description',
        currentStage: 'Test Product 2 current stage',
      },
    ];

    yield put(productsLoaded(products));
  } catch (err) {
    yield put(errorLoadingProducts(err));
  }
}

// Load a product's details
export function* loadProduct(action) {
  // const product = yield call(() => axios.get('products/' + action.payload).then(res =>
  //   res.data
  // ));
  const product = {
    name: 'Test Product 1',
    description: 'Test Description 1',
  };

  yield put(productLoaded(product));
}

// export function* createProduct(action) {
//   const requestURL = 'products';

//   try {
//     const products = yield call(request(requestURL, action.payload));
//     yield put(productsLoaded(products));
//   } catch (err) {
//     yield put(errorLoadingProducts(err));
//   }
// }

/**
 * Watcher sagas
 */

function* watchLoadProducts() {
  yield takeLatest(LOAD_PRODUCTS, loadProducts);
}

function* watchLoadProduct() {
  yield takeLatest(LOAD_PRODUCT, loadProduct);
}

function* watchCreateProduct() {
  yield takeLatest(CREATE_PRODUCT, createProduct);
}

export default function* rootSaga() {
  yield all([watchLoadProducts(), watchLoadProduct()]);
}
