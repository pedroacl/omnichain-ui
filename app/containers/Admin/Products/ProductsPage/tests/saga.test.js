/**
 * Tests for the ProductsPage sagas
 */
import { put } from 'redux-saga/effects';

import {
  loadProduct as loadProductAction,
  loadProducts as loadProductsAction,
  productLoaded as productLoadedAction,
  productsLoaded as productsLoadedAction,
  errorLoadingProducts as errorLoadingProductsAction,
} from 'containers/Admin/ProductsPage/actions';

import { loadProducts as loadProductsSaga, loadProduct as loadProductSaga } from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('loadProducts Saga', () => {
  let loadProductsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    loadProductsGenerator = loadProductsSaga(loadProductsAction(1));

    const selectDescriptor = loadProductsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = loadProductsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the productsLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'Test Product 1',
        description: 'Test Product 1 description',
      },
      {
        name: 'Test Product 2',
        description: 'Test Product 2 description',
      },
    ];

    const putDescriptor = loadProductsSaga(loadProductsAction(1));
    expect(putDescriptor.next().value).toEqual(put(productsLoadedAction(response)));
    expect(putDescriptor.next().done).toEqual(true);
  });

  // it('should call the productsLoadingError action if there is a response error', () => {
  //   const errMsg = 'Error loading products';
  //   const putDescriptor = loadProductsSaga(loadProductsAction(1)).throw(errMsg);
  //   expect(putDescriptor).toEqual(put(errorLoadingProductsAction(response)));
  // });
});

describe('loadProduct Saga', () => {
  let loadProductGenerator;

  beforeEach(() => {
    loadProductGenerator = loadProductSaga(loadProductAction(1));

    const selectDescriptor = loadProductGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = loadProductGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadProduct action if the request is successful', () => {
    const product = {
      name: 'Test Product 1',
      description: 'Test Description 1',
    };

    const putDescriptor = loadProductSaga(loadProductAction(1));
    expect(putDescriptor.next().value).toEqual(put(productLoadedAction(product)));
  });
});
