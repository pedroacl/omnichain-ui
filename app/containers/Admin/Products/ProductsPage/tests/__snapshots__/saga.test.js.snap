// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadProduct Saga should dispatch the loadProduct action if the request is successful 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "description": "Test Description 1",
        "name": "Test Product 1",
      },
      "type": undefined,
    },
    "channel": null,
  },
}
`;

exports[`loadProduct Saga should dispatch the loadProduct action if the request is successful 2`] = `undefined`;

exports[`loadProducts Saga should dispatch the productsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Array [
        Object {
          "description": "Test Product 1 description",
          "name": "Test Product 1",
        },
        Object {
          "description": "Test Product 2 description",
          "name": "Test Product 2",
        },
      ],
      "type": "omnichain/Admin/ProductPages/PRODUCTS_LOADED",
    },
    "channel": null,
  },
}
`;

exports[`loadProducts Saga should dispatch the productsLoaded action if it requests the data successfully 2`] = `undefined`;
